frontend/src/app/utils.ts

```diff
@@ -10,4 +17,29 @@ export function sortBy<T>(array: Array<T>, extractor: (T) => any) {
     }
     return 0;
   });
+  return result;
+}
+
+/**
+ * Takes a string containing placeholders such as ${foo}, and replaces them by their value in the given
+ * parameters. Beware: no expression supported, no space in curly braces.
+ *
+ * So interpolate('hello ${name}', { name: 'JB' }) returns "hello JB".
+ *
+ * This function is only used (currently) to generate functional error messages from code and parameters
+ * coming from the backend.
+ */
+export function interpolate(template: string, parameters: {[key: string]: any}): string {
+  // ugly loop because JS doesn't have a Regexp.quote() method, nor a replaceAll method. replace is supposed to replace
+  // all, but it does not.
+  let result = template;
```

Comment:
I found this surprising, but I suck at Regexp... You did not forget the global flag 'g' ?